src/CommandAndStateMessageParser.cpp

```diff
@@ -0,0 +1,317 @@
+#include "CommandAndStateMessageParser.hpp"
+#include <iostream>
+
+using namespace std;
+using namespace base;
+using namespace deep_trekker;
+
+CommandAndStateMessageParser::CommandAndStateMessageParser()
+    : mReader(mRBuilder.newCharReader())
+{
+}
+
+bool CommandAndStateMessageParser::parseJSONMessage(char const* data, string& errors)
+{
+    return mReader->parse(data, data + strlen(data), &mJData, &errors);
+}
+
+void CommandAndStateMessageParser::validateFieldPresent(
+    Json::Value const& value,
+    string const& fieldName
+)
+{
+    if (!value.isMember(fieldName))
+    {
+        throw invalid_argument("message does not contain the " + fieldName + " field");
+    }
+}
+
+bool CommandAndStateMessageParser::checkDeviceMacAddress(string address)
+{
+    return mJData["devices"].isMember(address);
+}
+
+string CommandAndStateMessageParser::parseGetMessage(string api_version)
```

Comment:
All the conversion methods should be unit tested on the C++ side. It would be faster, and it's closer to the actual implementation.

It would make the orogen side simpler: just write some port and make sure you get a JSON document of the right type. It would be fair to assume that in this case you went through the right codepath and are covered by the unit tests.